axon/networkbase.go

```diff
@@ -929,6 +932,50 @@ func (nt *NetworkBase) ReadWtsCpp(r io.Reader) error {
 	return err
 }
 
+// WtsSlice sets all the weights in recv order in given slice, resizing as needed
+func (nt *Network) WtsSlice(wts *[]float32) {
+	numSyns := 0
+	for _, ly := range nt.Layers {
+		for _, pj := range ly.RcvPrjns {
+			numSyns += len(pj.Syns)
+		}
+	}
+	if cap(*wts) >= numSyns {
+		*wts = (*wts)[:numSyns]
+	} else {
+		*wts = make([]float32, numSyns)
+	}
+	i := 0
+	for _, ly := range nt.Layers {
+		for _, pj := range ly.RcvPrjns {
+			for j := range pj.Syns {
+				(*wts)[i] = pj.Syns[j].Wt
+				i++
+			}
+		}
+	}
+}
+
+// WtsHash returns a hash code of all weight values
+func (nt *Network) WtsHash() string {
+	var wts []float32
+	nt.WtsSlice(&wts)
```

Comment:
Why this extra complexity playing around with the capacity of the slice, why not just allocate the slice inside WtsSlice? This function will be called rarely anyway, no need to think about perf here, plus that codepath `if cap(*wts) >= numSyns` never even triggers.

Reply:
This is just my policy for anything that involves allocating a potentially large slice -- do this for all the slices in etable etc -- it may never be used but anyway, I just follow the policy so I don't have to think about it! Definitely pays off in etable context.